federation/src/auth/JWE.ts

```diff
@@ -0,0 +1,103 @@
+} from 'sodium-native'
+
+const crypto_scalarmult_curve25519_BYTES = 32
+
+export const verifyToken = async (
+  token: string,
+  keyPair: { publicKey: Buffer; privateKey: Buffer },
+  nonce: number | null = null,
+) => {
+  const pubKeyX = Buffer.alloc(crypto_scalarmult_curve25519_BYTES)
+  const privKeyX = Buffer.alloc(crypto_scalarmult_curve25519_BYTES)
+  crypto_sign_ed25519_pk_to_curve25519(pubKeyX, keyPair.publicKey.subarray(0, 32))
+  crypto_sign_ed25519_sk_to_curve25519(privKeyX, keyPair.privateKey.subarray(0, 32))
+  const key = createPrivateKey({
+    key: {
+      kty: 'OKP',
+      crv: 'X25519',
+      x: pubKeyX.toString('base64url'),
+      d: privKeyX.toString('base64url'),
+    },
+    format: 'jwk',
+  })
+
+  const { plaintext } = await compactDecrypt(token, key)
+  const decryptedJWEPayload = JSON.parse(new TextDecoder().decode(plaintext)) as {
+    jws: string
+    publicKey: string
+  }
+  const foreignPub = createPublicKey({
+    key: {
+      kty: 'OKP',
+      crv: 'Ed25519',
+      x: Buffer.from(decryptedJWEPayload.publicKey).toString('base64url'),
+    },
+    format: 'jwk',
+  })
+  // TODO: verify clientPubKey (e.g. check federation table + rights here)
+  const { payload: payloadJWS } = await compactVerify(decryptedJWEPayload.jws, foreignPub)
+  const receivedNonce = JSON.parse(new TextDecoder().decode(payloadJWS)) as {
+    nonce: number
+    time: Date
+  }
+
+  if (nonce && receivedNonce.nonce !== nonce) {
+    throw new Error('Could not verify nonce')
+  }
+
+  return { publicKey: Buffer.from(decryptedJWEPayload.publicKey), nonce: receivedNonce.nonce }
+}
+
+export const generateToken = async (
+  nonce: number,
+  keyPair: { publicKey: Buffer; privateKey: Buffer },
+  receiverPublicKey: Buffer,
+) => {
+  const receiverPublicKeyFixed = receiverPublicKey // Buffer.from(receiverPublicKey.toString(), 'hex')
+  const receiverPublicKeyFixedX = Buffer.alloc(crypto_scalarmult_curve25519_BYTES)
+  crypto_sign_ed25519_pk_to_curve25519(receiverPublicKeyFixedX, receiverPublicKeyFixed)
+  const key = createPrivateKey({
+    key: {
+      kty: 'OKP',
+      crv: 'Ed25519',
+      x: keyPair.publicKey.toString('base64url'),
+      d: keyPair.privateKey.subarray(0, 32).toString('base64url'),
+    },
+    format: 'jwk',
+  })
+  const foreignPub = createPublicKey({
+    key: {
+      kty: 'OKP',
+      crv: 'X25519',
+      x: receiverPublicKeyFixedX.toString('base64url'),
+    },
+    format: 'jwk',
+  })
+  const jws = await new CompactSign(
+    new TextEncoder().encode(
+      JSON.stringify({
+        nonce,
+        time: new Date(),
+      }),
+    ),
+  )
+    .setProtectedHeader({ alg: 'EdDSA' })
+    .sign(key)
+  return await new CompactEncrypt(
+    new TextEncoder().encode(
+      JSON.stringify({
+        jws,
+        publicKey: keyPair.publicKey.subarray(0, 32),
+      }),
+    ),
+  )
+    .setProtectedHeader({ alg: 'ECDH-ES', enc: 'A256GCM' })
+    .encrypt(foreignPub)
+}
```

Comment:
If this file JWE.ts is the same as in backend/src/federation/auth, what do you think about the idea to create a kind of common modul, to insert there all general parts of code, which will be used from several moduls?
It allows us to keep the code only once, but use it multi in several moduls.
I think we still have or at least we will have in future more than this single file, which will be used from several moduls. One example for multi usage will be the DAO-layer classes, which could be usefull for backend- and federation-modul like it is for this JWE file too.

Reply:
I am a big fan of removing duplication, but we need to find a strategy to do this - its nothing we can just do - we need to decide an approach and do the corresponding refactoring.
I do not think it to be wise to do a rush job on a system which impacts alot of subsystems of ours.

For further read I'd like to link the following:
- https://dev.to/michalbryxi/share-common-code-with-yarn-workspaces-5g29

We have such a shared codebase with `database` with all its up and downsides.

Main problems I see are:
- Frontend and Backend use different javascript flavors (js, ts)
- Scope of a shared module must be well defined in order to prevent all packages to diffuse into the common module(s)
- Approach to use (workspaces, tsconfig includes, npm packages) must be chosen wisely based on shared package

We could make an epic for this - but I do not want to solve the problem in this PR.